### PR TITLE
[PULL REQUEST] Run directory updates enabling automated run directory creation for GCHP and GCClassic

### PR DESCRIPTION
**Pull request by LiamBindle**
Hi everyone,

This PR contains updates to run directory creation to facilitate automated run directory creation (and in general, run directory creation from a list of files with variables describing the state to initialize the run directory to). This PR is zero-diff on run directories created with `./createRunDir.sh`, but it does change the underlying way in which `./createRunDir.sh` works. 

In essence, `./createRunDir.sh` used to run a bunch of `sed` commands to tokens in the templates. This PR replaces this `sed`-based approach with the `envsubst`-based approach, and moves the copying/templating of *required* files to a new script called `init_rd.sh`. This new script has the following usage:

```console
$ init_rd.sh -h
Initialize the current working directory as a run directory (minimal file set). 

usage: 
   init_rd.sh FILE...
      FILE...: Files defining variables used to initialize run directory. The
               rightmost file has the highest precedence.
               
               Each FILE is sourced and the loaded variables are used to
               fill the run directory templates using envsubst.

   init_rd.sh --rdi-vars
      Lists all the RDI variables found in the templates.

   init_rd.sh [-h] [--help]
      Prints this help message.

```

In summary, `init_rd.sh` takes a list of files, loads the variables in those files, and then fills the templates (initializing the cwd as the run directory).

This PR updates `./createRunDir.sh` to use `init_rd.sh`. Note that `./createRunDir.sh` is still used to create all the supplemental files (e.g., runSampleScripts, etc.).

### Additional Notes
The variables that used to initialize run directory settings used to be quite vague (understandably since they were only used internally in `./createRunDir.sh`). In this PR I've updated the variables to be more descriptive and prefixed with `RDI_` (standing for **R**un **D**irectory **I**nitialization). This has the added benefit of allowing the `--rdi-vars` options which lists all the RDI variables found in the templates.

The main feature of this PR is `init_rd.sh`. I didn't intend for this script to be used by the public, but I do think it's useful to advanced users. It allows run directories to be created rapidly, precisely, and without the interactive wizard. This facilitates automated run directory creation. 

You can add the directory containing `init_rd.sh` to your `$PATH`. In doing this, you can run `init_rd.sh` from anywhere, and it will resolve the template locations correctly.

### Example
This example demonstrates creating a benchmark run directory (in a way that could be done by an automated system). First, we need a file with all the RDI variables:
```
RDI_DATA_ROOT=/ExtData
RDI_SIM_NAME=fullchem
RDI_MET="MERRA2"
RDI_MET_SOURCE=$RDI_MET

RDI_MET_RES="05x0625"
RDI_MET_NATIVE_RES="0.5x0.625"
RDI_MET_LAT_RES="05"
RDI_MET_LON_RES="0625"
RDI_MET_EXT="nc4"

RDI_MET_CN_YR=2015
RDI_MET_PRES_UNIT="Pa "
RDI_MET_PRES_SCALE="0.01"
RDI_MET_DUST_SF="3.86e-4"
RDI_MET_DIR=$RDI_DATA_ROOT/GEOS_0.5x0.625/MERRA2
RDI_RESTART_FILE='initial_GEOSChem_rst.c${CS_RES}'_fullchem.nc
RDI_HIST_TIME_AVG_DUR="7440000"
RDI_HIST_TIME_AVG_FREQ="7440000"
RDI_HIST_INST_DUR="7440000"
RDI_HIST_INST_FREQ="7440000"
RDI_HIST_MONTHLY_DIAG="1"

RDI_CS_RES="48"

RDI_NUM_CORES="48"
RDI_CORES_PER_NODE="24"
RDI_NUM_NODES="2"

RDI_SIM_START_DATE="20190701"
RDI_SIM_END_DATE="20190801"
RDI_SIM_START_TIME="000000"
RDI_SIM_END_TIME="000000"
RDI_SIM_DUR_YYYYMMDD="00000100"
RDI_SIM_DUR_HHmmSS="000000"
```

For this example, pretend this file is called `benchmark_rdi_vars.txt`. Then, I can create a benchmark run directory like so
```
$ mkdir benchmark-rundir
$ cd benchmark-rundir
$ init_rd.sh path/to/benchmark_rdi_vars.txt
```
This initializes the cwd as the run directory. Note that the RDI variables can be split across multiple files. Also note that RDI variables can be specified multiple times, in which case the rightmost file has the highest precedence.

The resulting run directory looks like so
```console
$ ls
CAP.rc	ChemDir  ExtData.rc  GCHP.rc  HcoDir  HEMCO_Config.rc  HEMCO_Diagn.rc  HISTORY.rc  input.geos  input.nml  MetDir  OutputDir  runConfig.sh
```

### Co-benefits

This also make run directory creation more modular. For example, RDI variables for simulation type {fullchem, TransportTracer, benchmark} are split into their own `.txt` files in the new `settings/` subdirectory, and the RDI variables for {MERRA2, GEOS-FP} are also in their own `.txt` files. This should ease run directory creation maintanance.

### Zero-diff tests

Check marks indicate tests that passed.

Simulation types:
- [x] Full chemistry sim
- [x] TransportTracer

Simulation options:
- [x] Standard
- [x] Benchmark
- [x] Complex SOA
- [x] Marine POA
- [x] Acid uptake on dust
- [x] TOMAS
- [x] APM
- [x] RRTMG

Meteorology source:
- [x] MERRA2
- [x] GEOS-FP

Since each of the three categories above control settings that are mutually exclusive, a single test can fulfill multiple
categories (i.e. each one needs to be tested at least once, but combinations don't need to be checked).

#### Additional tests

Now, when you run `./createRunDir.sh` it will create a file named `rdi_vars.sh` in the run directory. This file contains
the **R**un **D**irectory **I**nitialization variables that were used to create the run directory. Therefore, 
the essential run directory files must be reproducable with the following command:

```
envsubst_create.sh rdi_vars.txt
```

This test passed (i.e. all the essential run directory files were reproduced identically; checked with `diff`).

### CODE REVIEW

**Comment by LiamBindle** [comment]
Oops, I didn't realize that `alpha.10` was behind in the run directory. I'll work on rebasing on `dev/13.0.0`.

**Comment by lizziel** [comment]
Is this needed for 13.0.0? I thought we had decided it would go into 13.1, with updates for both GCClassic and GCHP at the same time.

**Comment by LiamBindle** [comment]
It's okay if this goes in after 13.0.0! There isn't a `dev/13.1.0` branch yet, so I put the target onto 13.0.0.

A mechanism like `init_rd.sh` is going to be extremely helpful to my timing tests/CS archive/transport tracers work, so I kind needed something like this anyways. I wanted to open the PR so that (1) it's available for my AIST work (via cherry-picking; and I'll be working on 13.0.0 prereleases), and (2) so that further run directory creation updates can build upon this.

**Comment by LiamBindle** [comment]
I cleaned up the commit log.

**Review by msulprizio — APPROVED** [review]

**Comment by msulprizio** [comment]
I just updated this branch to 13.1.0-alpha.2 and resolved conflicts. We still need to incorporate these changes into run/GCClassic for consistency between GCClassic and GCHP.

@lizziel Please review the latest changes before I bring into `dev`. I've added you as a reviewer.

@liam Do you have any objections to us removing the `RDI_` prefix from variables? Both Lizzie and I found it wasn't immediate what that stood for (without looking at the summary above).

File: run/GCHP/init_rd.sh [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+   echo "$DESCRIPTION"
+   exit 0
+fi 
+
+set -e
+
+THIS_SCRIPTS_DIRECTORY=$(realpath $(dirname "$0"))
+
+if [[ ( $* == --rdi-vars ) ]]; then 
+   grep -roh 'RDI_[A-Z_][A-Z_]*' $THIS_SCRIPTS_DIRECTORY | grep -v 'RDI_VARS' |  sort | uniq
```
Comment (LiamBindle) [inline run/GCHP/init_rd.sh:28]: This is the line that prints all the run directory initialization variables

**Comment by LiamBindle** [comment]
@msulprizio I think a unique prefix might be useful for tracking what templating variables exist in the template directories (and for differentiating them from things like shell variables in `runConfig.sh`). For example, if you run `init_rd.sh --rdi-vars` it prints a list of all the template variables. 

This way we don't need to keep a list of all the variables that are valid; instead we know it's any variable that start with `RDI_`. 

Would an alternative unique prefix be prefered? What do you think?

**Comment by msulprizio** [comment]
>Would an alternative unique prefix be preferred? What do you think?

Maybe. I'll leave it as is for now, and we can decide if we want to use a more descriptive prefix later.

**Comment by msulprizio** [comment]
@LiamBindle I'm working on expanding this capability for GCClassic, which involves incorporating more simulation types. I need to create a `*_settings.txt` file for each specialty simulation (e.g. mercury, CH4, POPs). Would you have any objections to moving the following settings out of the `fullchem_settings.txt`:

```
RDI_CS_RES="24"

RDI_NUM_CORES="24"
RDI_CORES_PER_NODE="24"
RDI_NUM_NODES="1"
```

I envision creating files for `c24_setting.txt`, `4x5_settings.txt`, etc. that could be selected based on whether the run directory is for GCClassic or GCHP.

**Comment by LiamBindle** [comment]
@msulprizio That's a good idea. Grouping those settings into resolution dependent groups makes sense I think.

**Comment by msulprizio** [comment]
>My only concern is the RUNDIR_VARS that are generated dynamically in createRunDir.sh. What do you think of moving those variables to static .txt files that are included based on the users input?

Yes, I had originally designed it that way, but found there were many overlapping settings between simulation types which led to duplicate variables in the output `rundir_vars.txt`. I can try to revisit this now that all of the GCClassic simulations have been sorted through and extract these from `createRunDirs.sh` again. I'll attempt this early next week and keep you posted.

**Comment by msulprizio** [comment]
> > My only concern is the RUNDIR_VARS that are generated dynamically in createRunDir.sh. What do you think of moving those variables to static .txt files that are included based on the users input?
> 
> Yes, I had originally designed it that way, but found there were many overlapping settings between simulation types which led to duplicate variables in the output `rundir_vars.txt`. I can try to revisit this now that all of the GCClassic simulations have been sorted through and extract these from `createRunDirs.sh` again. I'll attempt this early next week and keep you posted.

I was not able to address this point before going on maternity leave. I suggest merging as-is in 13.4.0 to avoid further delay on this PR and incurring additional conflicts. We can create a new feature request for this point. As noted above there is still room for other future work here. Some manipulation of run directory files is still being done in `createRunDir.sh`, `setupConfigFiles.sh`, and `intTestCreate.sh`, and this should ideally be incorporated in `init_rundir.sh` or the .txt files.

**Review by yantosca — APPROVED** [review]
This is a straightforward PR, it just involves text replacement in the run-directory scripts and configuration files.  OK to merge. 

One minor update, instead of using rm -f to free symbolic links, I would use unlink, which is perhaps safer.  I can make that change.

**Comment by yantosca** [comment]
The `feature/automated_rundir_creation` branch has been merged up to dev and stored in a new branch `dev+automated_rundir_creation`.  GCHP and GCC integration tests have been scheduled.

**Comment by yantosca** [comment]
I have merged the `feature/automated_rundir_creation` branch up to `dev` and stored it in a new branch (`dev+automatic_rundir_creation`) for testing.  The GCC integration tests all passed:

```console
fSummary of test results:
------------------------------------------------------------------------------
Execution tests passed: 80
Execution tests failed: 0
Execution tests not yet completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
but the GCHP tests all failed.  The gchp.log file was not created in each of the run directories.
```
==============================================================================
GCHP: Execution Test Results

Number of execution tests: 3
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gchp_fullchem_benchmark_merra2_c48...............Execute Simulation.....FAIL
gchp_fullchem_standard_merra2_c24................Execute Simulation.....FAIL
gchp_TransportTracers_geosfp_c24.................Execute Simulation.....FAIL
 
Summary of execution test results:
------------------------------------------------------------------------------
Execution tests passed:        0
Execution tests failed:        3
Execution tests not completed: 0
```
Still investigating ...

**Comment by yantosca** [comment]
Here is the `rundir_vars.txt` from the gchp_TransportTracers_merra2.c24` folder 
```
RUNDIR_ACID_UPTAKE='F'
RUNDIR_AERO_BCPI='* AERO_BCPI      $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_BCPI  2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_AERO_BCPO='* AERO_BCPO      $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_BCPO  2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_AERO_DST1='* AERO_DST1      $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_DST1  2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_AERO_NH4='* AERO_NH4       $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_NH4   2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_AERO_NIT='* AERO_NIT       $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_NIT   2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_AERO_OCPI='* AERO_OCPI      $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_OCPI  2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_AERO_OCPO='* AERO_OCPO      $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_OCPO  2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_AERO_SO4='* AERO_SO4       $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_SO4   2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_Br_GC='* Br_GC          $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_Br    2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_BrO_GC='* BrO_GC         $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_BrO   2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_BrO_TOMCAT='* BrO_TOMCAT     $ROOT/MERCURY/v2014-09/BrOx/BrOx.GMI.geos5.2x25.nc                                  LBRO2H                 1985/1-12/1/0 C xyz pptv     * - 1 1'
RUNDIR_Br_TOMCAT='* Br_TOMCAT      $ROOT/MERCURY/v2014-09/BrOx/BrOx.GMI.geos5.2x25.nc                                  LBRO2N                 1985/1-12/1/0 C xyz pptv     * - 1 1'
RUNDIR_Bry_DIR='STRAT/v2015-01/Bry'
RUNDIR_CH4_LOSS='* CH4_LOSS       $ROOT/CH4/v2014-09/4x5/gmi.ch4loss.geos5_47L.4x5.nc                                 CH4loss           1985/1-12/1/0      C xyz s-1      * - 1 1'
RUNDIR_CMIP6_FIELDS='false'
RUNDIR_CO2_COPROD='* CO2_COPROD     $ROOT/CO2/v2019-02/CHEM/CO2_prod_rates.GEOS5.2x25.47L.nc                            LCO               2004-2009/1-12/1/0 C xyz kgC/m3/s * - 1 1'
RUNDIR_COMPLEX_SOA='F'
RUNDIR_CORES_PER_NODE='48'
RUNDIR_CS_RES='48'
RUNDIR_DATA_ROOT=/n/holyscratch01/external_repos/GEOS-CHEM/gcgrid/data/ExtData
RUNDIR_DUSTALK_EXT='off'
RUNDIR_DUSTDEAD_EXT='off'
RUNDIR_GCAP2_VERTRES='47'
RUNDIR_GC_MODE='GCHP'
```

**Comment by yantosca** [comment]
I applied the fix in commit 43f90e7d5fb01037b1045f52188ed46f92a150ba, which fixed the issue.  The GCHP integration tests were failing because symbolic links to the various data directories had been removed inadvertently.  This is now fixed and the GCHP integration tests all pass:

```console
Summary of execution test results:
------------------------------------------------------------------------------
Execution tests passed:        3
Execution tests failed:        0
Execution tests not completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

**Pull request merged**